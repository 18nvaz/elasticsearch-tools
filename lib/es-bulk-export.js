#!/usr/bin/env node
'use strict';

var helpers = require('./helpers');
var fs = require('fs');
var program = require('commander');
var elasticsearch = require('elasticsearch');
var appInfo = require('../package.json');
var ProgressBar = require('progress');
var bar;

var esOptions = {
	index: 'a comma-separated list of index names to search; use _all or empty string to perform the operation on all indices',
	type: 'a comma-separated list of document types to search; leave empty to perform the operation on all types',
	body: 'the body to send along with this request.',
	analyzer: 'The analyzer to use for the query string',
	analyzeWildcard: 'specify whether wildcard and prefix queries should be analyzed (default: false)',
	fields: 'a comma-separated list of fields to return as part of a hit (default: "*")',
	from: 'starting offset (default: 0)',
	q: 'query in the Lucene query string syntax',
	routing: 'a comma-separated list of specific routing values',
	scroll: 'specify how long a consistent view of the index should be maintained for scrolled search (default: 1m)',
	size: 'number of hits to return',
	sort: 'a comma-separated list of <field>:<direction> pairs',
	timeout: 'explicit operation timeout'
};

// setup command line options
program
	.version(appInfo.version, '-v, --version')
	.option('-u, --url <url>', 'the elasticsearch url to connect to')
	.option('-f, --file <file>', 'the file to write data to');
// add es options
for (var key in esOptions) {
	if (esOptions.hasOwnProperty(key)) {
		program.option('--' + key + ' <' + key + '>', 'ES SEARCH OPTION: ' + esOptions[key]);
	}
}
// parse arguments
program.parse(process.argv);

// validate url and file
helpers.validateUrlAndFile(program);

// setup a default scroll value
if (!program.scroll) {
	program.scroll = '1m';
}

// init client
var client = new elasticsearch.Client({host: program.url});

// build our search object
var search = {
	fields: ['*'],
	_source: true,
	searchType: 'scan'
};
for (var key in esOptions) {
	if (esOptions.hasOwnProperty(key) && program.hasOwnProperty(key)) {
		search[key] = program[key];
	}
}

// declare our processing functions
var processed = 0;
var processResults = function (error, response) {
	var content = '', index, scrollOptions;
	if (error) {
		helpers.exit(error);
	}
	// init progress bar if needed
	if (!bar) {
		bar = new ProgressBar('processing :current of :total [:bar] :percent :elapseds', {
			width: 20,
			total: response.hits.total
		});
	}
	bar.render();
	// process results
	response.hits.hits.forEach(function (hit) {
		var meta = {index:{}};
		var doc = hit._source || {};
		var fields = hit.fields || {};
		// build meta
		for (var key in hit) {
			if (hit.hasOwnProperty(key) && key !== '_source' && key !== 'fields') {
				meta.index[key] = hit[key];
			}
		}
		for (var key in fields) {
			if (fields.hasOwnProperty(key)) {
				meta.index[key] = fields[key];
			}
		}
		content += JSON.stringify(meta) + '\n' + JSON.stringify(doc) + '\n';
		processed++;
		bar.tick();
	});
	fs.appendFileSync(program.file, content, 'utf-8');
	// continue to scroll
	if (response.hits.total !== processed) {
		scrollOptions = {scrollId: response._scroll_id};
		if (program.scroll) {
			scrollOptions.scroll = program.scroll;
		}
		client.scroll(scrollOptions, processResults);
	} else {
		bar.render();
		console.log('Done!');
		process.exit();
	}
};

// empty our file
fs.writeFileSync(program.file, '', 'utf-8');

// perform our search and start scrolling
client.search(search, processResults);

