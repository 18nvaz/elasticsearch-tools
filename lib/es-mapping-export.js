#!/usr/bin/env node
'use strict';

var helpers = require('./helpers');
var fs = require('fs');
var program = require('commander');
var elasticsearch = require('elasticsearch');
var appInfo = require('../package.json');

var esOptions = {
	index: 'String, String[], Boolean — A comma-separated list of index names',
	type: 'String, String[], Boolean — A comma-separated list of document types',
	ignoreUnavailable: 'Boolean — Whether specified concrete indices should be ignored when unavailable (missing or closed)',
	allowNoIndices: 'Boolean — Whether to ignore if a wildcard indices expression resolves into no concrete indices. (This includes _all string or when no indices have been specified)',
	expandWildcards: 'String — Whether to expand wildcard expression to concrete indices that are open, closed or both.',
	local: 'Boolean — Return local information, do not retrieve the state from master node (default: false)'
};

// setup command line options
program
	.version(appInfo.version, '-v, --version')
	.option('-u, --url <url>', 'the elasticsearch url to connect to')
	.option('-f, --file <file>', 'the file to write data to');
// add es options
for (var key in esOptions) {
	if (esOptions.hasOwnProperty(key)) {
		program.option('--' + key + ' <' + key + '>', 'ES SEARCH OPTION: ' + esOptions[key]);
	}
}
// parse arguments
program.parse(process.argv);

// validate url and file
helpers.validateUrlAndFile(program);

// init client
var client = new elasticsearch.Client({host: program.url});

// build our params object
var params = {};
for (var key in esOptions) {
	if (esOptions.hasOwnProperty(key) && program.hasOwnProperty(key)) {
		params[key] = program[key];
	}
}

// empty our file
fs.writeFileSync(program.file, '', 'utf-8');

// get mapping, and write to file
client.indices.getMapping(params, function (error, response) {
	if (error) {
		helpers.exit(error);
	}
	var keys = Object.keys(response);
	fs.appendFileSync(program.file, JSON.stringify(response, null, '  '), 'utf-8');
	console.log('Exporting mappings for: ', keys);
	console.log('Done!');
	process.exit();
});
