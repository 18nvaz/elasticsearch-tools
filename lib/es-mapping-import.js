#!/usr/bin/env node
'use strict';

var helpers = require('./helpers');
var fs = require('fs');
var program = require('commander');
var elasticsearch = require('elasticsearch');
var appInfo = require('../package.json');
var async = require('async');

var esOptions = {
	ignoreConflicts: 'Boolean — Specify whether to ignore conflicts while updating the mapping (default: false)',
	timeout: 'Date, Number — Explicit operation timeout',
	masterTimeout: 'Date, Number — Specify timeout for connection to master',
	ignoreUnavailable: 'Boolean — Whether specified concrete indices should be ignored when unavailable (missing or closed)',
	allowNoIndices: 'Boolean — Whether to ignore if a wildcard indices expression resolves into no concrete indices. (This includes _all string or when no indices have been specified)',
	expandWildcards: 'String — Whether to expand wildcard expression to concrete indices that are open, closed or both.'
};

// setup command line options
program
	.version(appInfo.version, '-v, --version')
	.option('-u, --url <url>', 'the elasticsearch url to connect to')
	.option('-f, --file <file>', 'the file to write data to');
// add es options
for (var key in esOptions) {
	if (esOptions.hasOwnProperty(key)) {
		program.option('--' + key + ' <' + key + '>', 'ES SEARCH OPTION: ' + esOptions[key]);
	}
}
// parse arguments
program.parse(process.argv);

// validate url and file
helpers.validateUrlAndFile(program);

// init client
var client = new elasticsearch.Client({host: program.url});

// build our params object
var params = {};
for (var key in esOptions) {
	if (esOptions.hasOwnProperty(key) && program.hasOwnProperty(key)) {
		params[key] = program[key];
	}
}

// read file, and create indexes/mappings
fs.readFile(program.file, 'utf-8', function (err, contents) {
	if (err) {
		helpers.exit(err);
	}
	try {
		var parsedContent = JSON.parse(contents);
	} catch (err) {
		helpers.exit('Cannot parse file: ' + err.toString());
	}
	var indexes = Object.keys(parsedContent);
	var mappings = [];
	var created = [];
	// build our mappings array
	indexes.forEach(function (index) {
		var m = parsedContent[index].mappings;
		// loop through types
		Object.keys(m).forEach(function (type) {
			mappings.push({
				index: index,
				type: type,
				body: m[type]
			});
		});
	});
	console.log('Creating Indexes (if needed):');
	async.eachSeries(mappings, function (mapping, callback) {
		var prefix = '  [index="' + mapping.index + '"]';
		if (created.indexOf(mapping.index) >= 0) {
			callback();
		} else {
			created.push(mapping.index);
			client.indices.exists({index: mapping.index}, function (error, response) {
				if (err) {
					callback(err);
				} else if (response) {
					console.log(prefix + ' - exists');
					callback();
				} else {
					console.log(prefix + ' - creating');
					client.indices.create({index: mapping.index}, callback);
				}
			});
		}
	}, function (err) {
		if (err) {
			helpers.exit(err);
		}
		console.log('Creating Mappings:');
		async.eachSeries(mappings, function (mapping, callback) {
			// put mapping
			console.log('  [index="' + mapping.index + '",type="' + mapping.type + '"]');
			params.index = mapping.index;
			params.type = mapping.type;
			params.body = mapping.body;
			client.indices.putMapping(params, callback);
		}, function (err) {
			if (err) {
				helpers.exit(err);
			}
			console.log('Done!');
			process.exit();
		});
	});
});
