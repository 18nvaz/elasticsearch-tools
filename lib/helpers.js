// a few helper functions

exports.exit = function (str) {
	str = str.toString().replace(/^error:/i, '');
	console.error('  error: ' + str);
	process.exit(1);
};

exports.validateUrlAndFile = function (program) {
	['url', 'file'].forEach(function (key) {
		if (typeof program[key] !== 'string') {
			exports.exit('You must pass in a valid --' + key + ' option.');
		}
	});
};
