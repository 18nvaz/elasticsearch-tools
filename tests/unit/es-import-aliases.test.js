jest.mock('../../lib/importer.js');

test('es-import-alises', () => {
  const importerMock = require('../../lib/importer');
  require('../../lib/es-import-aliases');
  expect(importerMock.indicesImport.mock.calls).toMatchSnapshot();
});
