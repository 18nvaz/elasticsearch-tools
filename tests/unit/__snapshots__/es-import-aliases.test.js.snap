// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`es-import-alises 1`] = `
Array [
  Array [
    "putAlias",
    "aliases",
    "name",
    Object {
      "masterTimeout": "Date, Number — Specify timeout for connection to master",
      "timeout": "Date, Number — Explicit operation timeout",
    },
  ],
]
`;
